### PR TITLE
Introduce an ngeox.profile.ProfileOptions

### PR DESCRIPTION
**Pull request by gberaudo**
An `ngeox.profile.ProfileOptions` typedef has been introduced and must be passed to the `ngeo.profile` factory. Additionally, the profile now depends on an `ngeo.ProfileExtractor`  to support any tabular data.

The directive has been updated to require the options attribute.
Basic support for updating the profile was added.

### CODE REVIEW

File: externs/ngeox.js [hunk 1]
```diff
@@ -49,3 +49,33 @@ ngeox.interaction.MeasureOptions.prototype.style;
...
+ * Options for the profile.
+ *  @typedef {{
+ *    extractor: ngeox.profile.ProfileExtractor,
+ *    light: (boolean|undefined),
+ *    hoverCallback: (function(Object)|undefined),
+ *    outCallback: (function()|undefined),
+ *    zKeys: (Array.<string>|undefined)
+ *  }}
+ */
+ngeox.profile.ProfileOptions;
```
Comment (elemoine) [inline externs/ngeox.js]: It's not sufficient for the externs definitions. We need to define a prototype and prototype properties. See MeasureOptions for an example.

File: src/d3-ext/profile.js [hunk 2]
```diff
@@ -58,13 +62,13 @@ ngeo.profile = function() {
    * @type {function(Object)}
    * Hover callback function.
    */
-  var hoverCallback = goog.nullFunction;
+  var hoverCallback = options.hoverCallback || goog.nullFunction;
```
Comment (elemoine) [inline src/d3-ext/profile.js]: Please use `goog.isDef` for consistency reasons.

Comment (gberaudo) [inline src/d3-ext/profile.js]: Using `goog.isDef` would be less readable, is it really an improvement?
`var hoverCallback = options.hoverCallback || goog.nullFunction;`
> var hoverCallback = goog.isDef(options.hoverCallback) ?  options.hoverCallback :
>   || goog.nullFunction;

Wouldn't it make sense to allow the short syntax when dealing with objects?

File: externs/ngeox.js (hunk 1 above)
Comment (gberaudo) [inline externs/ngeox.js]: I thought it was only a requirement of the documentation generation tool used by Ol3.
Where is defined `ngeox.interaction.MeasureOptions.prototype`?

File: src/d3-ext/profile.js (hunk 2 above)
Comment (elemoine) [inline src/d3-ext/profile.js]: I think we want to be consistent and use `goog.isDef` everywhere. Much easier to explain to new developers.

File: externs/ngeox.js (hunk 1 above)
Comment (elemoine) [inline externs/ngeox.js]: > I thought it was only a requirement of the documentation generation tool used by Ol3.

No.



> Where is defined ngeox.interaction.MeasureOptions.prototype?

Each measure option is defined in the prototype. E.g. https://github.com/camptocamp/ngeo/blob/master/externs/ngeox.js#L29. That is the way externs work.

File: src/d3-ext/profile.js (hunk 2 above)
Comment (gberaudo) [inline src/d3-ext/profile.js]: Consistency should be second after readability. There is no point in having a code base consistently hard to read.
Moreover, it is just plain and widely used javascript every developer should be familiar with.

Comment (elemoine) [inline src/d3-ext/profile.js]: I disagree.

1. I don't think

 ```js
   var hoverCallback = goog.isDef(options.hoverCallback) ? options.hoverCallback : goog.nullFunction;
 ```

 is harder to read than

 ```js
   var hoverCallback = options.hoverCallback || goog.nullFunction;
 ```

 The first expression is even more explicit, because it makes it clear that the test is about whether `hoverCallback` is defined in the `options` object.

2. It's easier to explain to people that they should **always** use `goog.isDef(options.theOption)` to test if `theOption` is defined in the options.

 I don't want to have to tell people that they can use `||` when dealing with objects, and `goog.isDef` when dealing with number, strings and booleans. I prefer systematic approaches.

Comment (gberaudo) [inline src/d3-ext/profile.js]: Honestly, having `options.hoverCallback` repeated makes it harder to read and easy to make typos.

While both expressions are not synonyms, they are both doing their job (I do not think we should use
goog.isDefAndNotNull).

I disagree with you on the second point: I expect all developers working on Ngeo to understand basics concepts of javascript. Then it is just a matter of having read the contributor guide. Don't you think preventing developers to use basic javascript is a bit tyrannical? ;-)

Comment (elemoine) [inline src/d3-ext/profile.js]: > Honestly, having options.hoverCallback repeated makes it harder to read and easy to make typos.

Come on! You really can't say that `var foo = goog.isDef(options.foo) ? options.foo : "default"` is hard to read.

> I disagree with you on the second point: I expect all developers working on Ngeo to understand basics concepts of javascript.

People very often make mistakes because `undefined`, `null`, `0` and `''` are all falsy values. People who see `var callback = options.callback || goog.nullFunction` will do the same for their options of type `number`: `var myNum = options.myNum || -1`.

I think it's a good thing to teach people to be stricter and more explicit with types.

**Comment by gberaudo** [comment]
@elemoine, I updated the code according to your two comments and it is now ready for merging.

File: src/d3-ext/profile.js (hunk 2 above)
Comment (gberaudo) [inline src/d3-ext/profile.js]: @elemoine, I think rules should be flexible enough so that they do not hinder developer creativity. Nevertheless, I updated the PR with your requirements.

File: examples/profile.js [hunk 3]
```diff
@@ -25,22 +25,21 @@ app.MainController = function($http, $scope) {
...
- */
-app.MainController.prototype.onProfileHover = function(point) {
-  this['point'] = point;
-};
-
-
-/**
- */
-app.MainController.prototype.onProfileOut = function() {
+  var profileOptions = {
```
Comment (elemoine) [inline examples/profile.js]: This object does not have all the properties defined in the `ngeox.profile.ProfileOptions` type alias, so it may make sense to cast the right value.

File: externs/ngeox.js [hunk 4]
```diff
@@ -49,3 +49,75 @@ ngeox.interaction.MeasureOptions.prototype.style;
...
+ * @type {Array.<string>|undefined}
+ */
+ngeox.profile.ProfileOptions.prototype.zKeys;
+
+
+/**
+ * Profile data extractor.
+ * @typedef {{
+ *   z: (function(Object, string=): number),
+ *   dist: (function(Object): number)
```
Comment (elemoine) [inline externs/ngeox.js]: Do you really need brackets here?

File: src/directives/profile.js [hunk 5]
```diff
@@ -27,31 +27,37 @@ ngeo.profileDirective = function() {
...
+          }
+
           var selection = d3.select(element[0]);
-          var profile = ngeo.profile();
+          var profile, options, data;
 
-          scope.$watch(attrs['ngeoProfile'], function(newVal, oldVal) {
-            if (goog.isDef(newVal)) {
-              selection.datum(newVal).call(profile);
+          scope.$watch(attrs['ngeoProfileOptions'], function(newVal, oldVal) {
```
Comment (elemoine) [inline src/directives/profile.js]: Why do you watch the options?

File: src/directives/profile.js [hunk 6]
```diff
@@ -27,31 +27,37 @@ ngeo.profileDirective = function() {
...
+              refreshData();
             }
           });
 
-          profile.onHover(function(point) {
-            scope.$apply(function() {
-              scope.$eval(attrs['ngeoProfileOnhover'], {'point': point});
-            });
-          });
+          function refreshData() {
```
Comment (elemoine) [inline src/directives/profile.js]: It would make more sense to define the function after the two watcher registrations.

File: src/directives/profile.js [hunk 7]
```diff
@@ -27,31 +27,37 @@ ngeo.profileDirective = function() {
...
+       * @param {angular.Scope} scope Scope.
+       * @param {angular.JQLite} element Element.
+       * @param {angular.Attributes} attrs Attributes.
+       */
         function(scope, element, attrs) {
 
+          var optionsAttr = attrs['ngeoProfileOptions'];
+          if (!goog.isDef(optionsAttr)) {
+            throw new Error('The ngeoProfileOptions attribute is required');
+          }
```
Comment (elemoine) [inline src/directives/profile.js]: We just use `goog.asserts.assert` elsewhere. In this case:

```js
goog.asserts.assert(goog.isDef(optionsAttr);
```

**Comment by elemoine** [comment]
The docs at the head of the `src/directives/profile.js` need to be updated.

File: examples/profile.html [hunk 8]
```diff
@@ -19,8 +19,8 @@
   </head>
   <body ng-controller="MainController as ctrl">
     <div class="profile" ngeo-profile="::ctrl.profileData"
-         ngeo-profile-onhover="ctrl.onProfileHover(point)"
-         ngeo-profile-onout="ctrl.onProfileOut()"></div>
+         ngeo-profile-options="::ctrl.profileOptions">
```
Comment (elemoine) [inline examples/profile.html]: You can remove the one-time binding expression here if you remove the "options" watcher.

File: src/directives/profile.js [hunk 9]
```diff
@@ -27,31 +27,37 @@ ngeo.profileDirective = function() {
...
     link:
         /**
-         * @param {angular.Scope} scope Scope.
-         * @param {angular.JQLite} element Element.
-         * @param {angular.Attributes} attrs Attributes.
-         */
+       * @param {angular.Scope} scope Scope.
+       * @param {angular.JQLite} element Element.
+       * @param {angular.Attributes} attrs Attributes.
+       */
```
Comment (elemoine) [inline src/directives/profile.js]: Why did you change the indentation? It looked correct to me.

**Comment by elemoine** [comment]
I've added a few other comments. If you agree with my comments, please merge when they are addressed.

File: externs/ngeox.js (hunk 4 above)
Comment (pgiraud) [inline externs/ngeox.js]: Why is the signature for `z` different?

File: examples/profile.js [hunk 10]
```diff
@@ -25,22 +25,58 @@ app.MainController = function($http, $scope) {
...
+
+  /**
+   * @type {ngeox.profile.ProfileExtractor}
+   */
+  var extractor = {z: z, dist: dist};
+
+
+  // Using closures for hoverCallback and outCallback since
+  // wrapping in angular.bind leads to a closure error.
+  var that = this;
```
Comment (pgiraud) [inline examples/profile.js]: Can you tell us more about this closure error?

File: examples/profile.js [hunk 11]
```diff
@@ -25,22 +25,58 @@ app.MainController = function($http, $scope) {
...
+
+
+  /**
+   * @type {ngeox.profile.ProfileOptions}
+   */
+  var profileOptions = {
+    extractor: extractor,
+    hoverCallback: hoverCallback,
+    light: false, // explicitely set to avoid closure error
+    zKeys: undefined, // explicitely set to avoid closure error
```
Comment (pgiraud) [inline examples/profile.js]: What is this `zKeys` option for?

File: examples/profile.js (hunk 10 above)
Comment (gberaudo) [inline examples/profile.js]: ERR! compile .build/examples/all.js:1547: ERROR - initializing variable
ERR! compile found   : {extractor: {dist: function ((Object|null)): number, z: function ((Object|null), string=): number}, hoverCallback: (Function|null), light: boolean, outCallback: function (): undefined, ...}
ERR! compile required: {extractor: {dist: function ((Object|null)): number, z: function ((Object|null), string=): number}, hoverCallback: (function ((Object|null)): ?|undefined), light: (boolean|undefined), outCallback: (function (): ?|undefined), ...}
ERR! compile   var profileOptions = {

File: examples/profile.js (hunk 11 above)
Comment (pgiraud) [inline examples/profile.js]: It's a bit unfortunate that we need to explicitely have to set values for options for which the default value is OK.

File: examples/profile.js (hunk 10 above)
Comment (gberaudo) [inline examples/profile.js]: `angular.bind` is defined to get a `Function` and return a `Function` so it actually loses the type of the function. Where is angular.bind jsdocs coming from?

File: examples/profile.js (hunk 11 above)
Comment (gberaudo) [inline examples/profile.js]: It is not used yet, I will remove it.

File: examples/profile.js (hunk 10 above)
Comment (elemoine) [inline examples/profile.js]: Look at `.build/externs/angular-1.3.js` for the Angular-related types.

Comment (gberaudo) [inline examples/profile.js]: I see `Copyright 2012 The Closure Compiler Authors.`. Where was it downloaded from?

Comment (elemoine) [inline examples/profile.js]: https://github.com/google/closure-compiler/blob/master/contrib/externs/angular-1.3.js

Comment (gberaudo) [inline examples/profile.js]: OK, I found it in the Makefile.

Comment (gberaudo) [inline examples/profile.js]: I will try fixing the jsdoc.

File: examples/profile.js [hunk 12]
```diff
@@ -25,22 +25,57 @@ app.MainController = function($http, $scope) {
...
+  };
+
+
+  /**
+   * @type {ngeox.profile.ProfileOptions}
+   */
+  var profileOptions = {
+    extractor: extractor,
+    hoverCallback: hoverCallback,
+    light: false, // explicitely set to avoid closure error
```
Comment (pgiraud) [inline examples/profile.js]: Once again, can't we use a default value? Does it really have to be explicitely set?

File: examples/profile.js (hunk 10 above)
Comment (gberaudo) [inline examples/profile.js]: OK, I have a fix, I will send it upstream.

File: examples/profile.js (hunk 11 above)
Comment (gberaudo) [inline examples/profile.js]: Indeed. @elemoine do you know of a better handling of this issue?
This case is similar to option typedefs we pass to constructors. Ex:
https://github.com/camptocamp/ngeo/blob/master/examples/geolocation.js#L36
```js
var map = new ol.Map({
  layers: [
    new ol.layer.Tile({
      source: new ol.source.OSM()
    })
  ],
  view: view
});
```

Comment (elemoine) [inline examples/profile.js]: On Tue, Mar 3, 2015 at 9:24 AM, gberaudo <notifications@github.com> wrote:
>
> In examples/profile.js:
>
> > +    that['point'] = point;
> > +  };
> > +
> > +  var outCallback = function() {
> > +    that['point'] = null;
> > +  };
> > +
> > +
> > +  /**
> > +   * @type {ngeox.profile.ProfileOptions}
> > +   */
> > +  var profileOptions = {
> > +    extractor: extractor,
> > +    hoverCallback: hoverCallback,
> > +    light: false, // explicitely set to avoid closure error
> > +    zKeys: undefined, // explicitely set to avoid closure error
>
> Indeed. @elemoine do you know of a better handling of this issue?
> This case is similar to option typedefs we pass to constructors. Ex:
> https://github.com/camptocamp/ngeo/blob/master/examples/geolocation.js#L36
>
> var map = new ol.Map({
>   layers: [
>     new ol.layer.Tile({
>       source: new ol.source.OSM()
>     })
>   ],
>   view: view
> });


You can use a typecast:


var profileOptions = /** @type {ngeox.profile.ProfileOptions} */ ({
  extractor: extractor,
  hoverCallback: hoverCallback
});


-- 
Eric Lemoine

Camptocamp France SAS
Savoie Technolac, BP 352
73377 Le Bourget du Lac, Cedex

Tel : 00 33 4 79 44 44 94
Mail : eric.lemoine@camptocamp.com
http://www.camptocamp.com

File: examples/profile.js (hunk 10 above)
Comment (gberaudo) [inline examples/profile.js]: See https://github.com/google/closure-compiler/pull/867

Comment (elemoine) [inline examples/profile.js]: Our applications are compiled and use `goog.bind` instead of `angular.bind`, so this is not a major problem for us.

File: externs/ngeox.js (hunk 4 above)
Comment (gberaudo) [inline externs/ngeox.js]: You are right, fixed.
I added this and zKeys in anticipation of multiple elevation values
handling. However, it is better if you add it when you actually implement
it.

File: examples/profile.js (hunk 10 above)
Comment (gberaudo) [inline examples/profile.js]: Can we use `goog.bind` in examples?

Comment (elemoine) [inline examples/profile.js]: No. https://github.com/camptocamp/ngeo/blob/master/docs/developer-guide.md#authoring-examples.

**Comment by gberaudo** [comment]
@elemoine, @pgiraud, does the current code is OK for you?
- use angular.bind with modified externs;
- directly set the ProfileOptions in this.

File: Makefile [hunk 13]
```diff
@@ -233,7 +233,7 @@ node_modules/angular/angular.min.js node_modules/angular-animate/angular-animate
 
 .build/externs/angular-1.3.js:
 	mkdir -p $(dir $@)
-	wget -O $@ https://raw.githubusercontent.com/google/closure-compiler/master/contrib/externs/angular-1.3.js
+	wget -O $@ https://github.com/gberaudo/closure-compiler/raw/master/contrib/externs/angular-1.3.js
```
Comment (elemoine) [inline Makefile]: I'd rather use a workaround in the example, because we don't know whether your patch will be merged.

**Comment by pgiraud** [comment]
@gberaudo we should avoid using forks as much as possible. We should open a new issue to keep track of the pull request in google closure.

Otherwise, it looks good to me.

File: Makefile (hunk 13 above)
Comment (gberaudo) [inline Makefile]: In this case, let's use the initial workaround I proposed and we will change if my PR to closure is merged.

Comment (pgiraud) [inline Makefile]: +1

**Comment by elemoine** [comment]
+1 for using the initial workaround for `angular.bind`.

**Comment by gberaudo** [comment]
I think I have addressed all comments and will now merge. Thank you @elemoine and @pgiraud for the review.

**Comment by gberaudo** [comment]
@elemoine, I do not have write access to the ngeo repository. Could you please add me to the project collaborators or merge this PR? Thank you.

**Comment by elemoine** [comment]
> @elemoine, I do not have write access to the ngeo repository. Could you please add me to the project collaborators or merge this PR? Thank you.

Done.

**Comment by gberaudo** [comment]
Thanks.
